### PR TITLE
Forward port changes from `iron` to `master` branch + Add new `rmw_count_*` functions 

### LINKED ISSUE #100: Runtime error message when publishing message (ros2 rolling / Iron )
Hi,

After the fix of #94 `rmw_iceoryx_cpp` can be compiled on Windows. Then I did some test and find an issue.

steps:

1) start iox-roudi.exe

2) `set RMW_IMPLEMENTATION=rmw_iceoryx_cpp` and then run `ros2 run demo_nodes_cpp talker`

then I have the following error messages:

```
'failed to resolve symbol 'rmw_get_gid_for_client' in shared library 'c:\ros\rolling\install\bin\rmw_iceoryx_cpp.dll', at C:\ros\rolling\src\ros2\rmw_implementation\rmw_implementation\src\functions.cpp:154'

'failed to resolve symbol 'rmw_take_dynamic_message' in shared library 'c:\ros\rolling\install\bin\rmw_iceoryx_cpp.dll', at C:\ros\rolling\src\ros2\rmw_implementation\rmw_implementation\src\functions.cpp:154'

'failed to resolve symbol 'rmw_take_dynamic_message_with_info' in shared library 'c:\ros\rolling\install\bin\rmw_iceoryx_cpp.dll', at C:\ros\rolling\src\ros2\rmw_implementation\rmw_implementation\src\functions.cpp:154'

'failed to resolve symbol 'rmw_serialization_support_init' in shared library 'c:\ros\rolling\install\bin\rmw_iceoryx_cpp.dll', at C:\ros\rolling\src\ros2\rmw_implementation\rmw_implementation\src\functions.cpp:154'
```

As for `rmw_get_gid_for_client`, it's added in ROS2 `Iron` (search "rmw_get_gid_for_client" at https://docs.ros.org/en/iron/Releases/Iron-Irwini-Complete-Changelog.html), and I'm testing on ros2 `rolling`, so it might be fine in `humble` but not in `Iron` when it's planned to be supported in #99 

BTW, to my surprise, the talker/listener demo still work great despite of the error message. 💯

### PR DESCRIPTION
**Pull request by mossmaurice**
## Todo

1. [x] Add `rmw_count_clients`
2. [x] Add `rmw_count_services`

Closes #100

### CODE REVIEW

File: rmw_iceoryx_cpp/src/rmw_count.cpp [hunk 1]
```diff
@@ -84,6 +84,46 @@ rmw_count_subscribers(
...
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(service_name, RMW_RET_ERROR);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(count, RMW_RET_ERROR);
+
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    rmw_count_services
+    : node, node->implementation_identifier,
+    rmw_get_implementation_identifier(), return RMW_RET_ERROR);
+
+  /// @todo There is no API to find out which 'ServiceDescription' is offered by which node
+  RMW_SET_ERROR_MSG("rmw_count_services is not supported in iceoryx");
```
Comment (budrus) [inline rmw_iceoryx_cpp/src/rmw_count.cpp:103]: @mossmaurice Hmm, you can provide the node name with the client and server options. Do we have the problem that this needs to be included in the introspection topics? For publisher and subscriber, this is available, right?

Comment (mossmaurice) [reply rmw_iceoryx_cpp/src/rmw_count.cpp:103]: @budrus I don't think that's possible. How would one access the server options here in this function? 

> Do we have the problem that this needs to be included in the introspection topics?

Correct, there are currently no introspection topics available for client/server. Hence, for the ROS graph I use `service_discovery.findService(..)` and [provide a dummy name](https://github.com/ros2/rmw_iceoryx/blob/iron/rmw_iceoryx_cpp/src/internal/iceoryx_topic_names_and_types.cpp#L159-L180).

> For publisher and subscriber, this is available, right?

:heavy_check_mark: 

Two options would be possible:

1. Add client/server introspection topics
2. Extend `iox::runtime::ServiceDiscovery` to return us info about the node association

Comment (budrus) [reply rmw_iceoryx_cpp/src/rmw_count.cpp:103]: @mossmaurice OK. 1 would be consistent with how we do it for pub/sub.

**Review by budrus — APPROVED** [review]

File: rmw_iceoryx_cpp/src/rmw_count.cpp (hunk 1 above)
Comment (mossmaurice) [reply rmw_iceoryx_cpp/src/rmw_count.cpp:103]: @budrus I agree, we already have an issue for that https://github.com/eclipse-iceoryx/iceoryx/issues/1128

**Pull request merged**